### PR TITLE
Optimize S3Escaper.encode()

### PR DESCRIPTION
**Pull request by kzhsw**
Replace `String.replaceAll` with StringBuilder for better performance.
See also benchmark at <https://github.com/kzhsw/minio-java-jmh>.
Close https://github.com/minio/minio-java/issues/1528

### CODE REVIEW

**Comment by harshavardhana** [comment]
> See also benchmark at https://github.com/kzhsw/minio-java-jmh.

where are the before and after results, do not see any results shared - please share them.

**Comment by kzhsw** [comment]
> > See also benchmark at https://github.com/kzhsw/minio-java-jmh.
> 
> where are the before and after results, do not see any results shared - please share them.

It's [here](https://github.com/kzhsw/minio-java-jmh/actions/runs/7752407466/job/21141853765#step:5:98)

File: api/src/main/java/io/minio/S3Escaper.java [hunk 1]
```diff
@@ -28,24 +28,32 @@ public static String encode(String str) {
...
-        .replaceAll("\\:", "%3A")
-        .replaceAll("\\;", "%3B")
-        .replaceAll("\\=", "%3D")
-        .replaceAll("\\@", "%40")
-        .replaceAll("\\[", "%5B")
-        .replaceAll("\\]", "%5D");
+    str = ESCAPER.escape(str);
+    StringBuilder encoded = new StringBuilder(str.length() + Math.max(str.length() >> 3, 4));
+
+    for (char ch : str.toCharArray()) {
```
Comment (balamurugana) [inline api/src/main/java/io/minio/S3Escaper.java:34]:
```suggestion
    StringBuilder encoded = new StringBuilder();
    for (char ch : ESCAPER.escape(str).toCharArray()) {
```

File: api/src/main/java/io/minio/S3Escaper.java [hunk 2]
```diff
@@ -28,24 +28,30 @@ public static String encode(String str) {
...
+        case ';': encoded.append("%3B"); break;
+        case '=': encoded.append("%3D"); break;
+        case '@': encoded.append("%40"); break;
+        case '[': encoded.append("%5B"); break;
+        case ']': encoded.append("%5D"); break;
+        default: encoded.append(ch);
+      }
+    }
+
+    return encoded.toString();
```
Comment (balamurugana) [inline api/src/main/java/io/minio/S3Escaper.java:54]:
```suggestion
    StringBuilder builder = new StringBuilder();
    for (char ch : ESCAPER.escape(str).toCharArray()) {
      switch (ch) {
        case '!':
          builder.append("%21");
          break;
        case '$':
          builder.append("%24");
          break;
        case '&':
          builder.append("%26");
          break;
        case '\'':
          builder.append("%27");
          break;
        case '(':
          builder.append("%28");
          break;
        case ')':
          builder.append("%29");
          break;
        case '*':
          builder.append("%2A");
          break;
        case '+':
          builder.append("%2B");
          break;
        case ',':
          builder.append("%2C");
          break;
        case '/':
          builder.append("%2F");
          break;
        case ':':
          builder.append("%3A");
          break;
        case ';':
          builder.append("%3B");
          break;
        case '=':
          builder.append("%3D");
          break;
        case '@':
          builder.append("%40");
          break;
        case '[':
          builder.append("%5B");
          break;
        case ']':
          builder.append("%5D");
          break;
        default:
          builder.append(ch);
      }
    }
    return builder.toString();
```

**Review by balamurugana — APPROVED** [review]

**Review by minio-trusted — APPROVED** [review]

**Pull request merged**